tests/test_spawner.py

```diff
@@ -520,6 +520,85 @@ async def test_spawn_progress(kube_ns, kube_client, config, hub_pod, hub):
     await spawner.stop()
 
 
+async def test_spawn_start_restore_pod_name(
+    kube_ns,
+    kube_client,
+    config,
+    hub,
+    exec_python,
+):
+    # Emulate stopping Jupyterhub and starting with different settings while some pods still exist
+
+    # Save state with old config
+    old_spawner = KubeSpawner(
+        hub=hub,
+        user=MockUser(name="start"),
+        config=config,
+        api_token="abc123",
+        oauth_client_id="unused",
+    )
+    old_spawner_pod_name = old_spawner.pod_name
+
+    # save state
+    await old_spawner.start()
+    old_state = old_spawner.get_state()
+    await old_spawner.stop()
+
+    # Change config
+    config.KubeSpawner.pod_name_template = 'new-{username}--{servername}'
+
+    # Load old state
+    spawner = KubeSpawner(
+        hub=hub,
+        user=MockUser(name="start"),
+        config=config,
+        api_token="abc123",
+        oauth_client_id="unused",
+    )
+    spawner.load_state(old_state)
```

Comment:
Is everything above representing the emulation of JupyterHub being stopped, configured, and started?

Reply:
`user.spawn()` calls `Spawner.get_state()` and saves it to the database:
https://github.com/jupyterhub/jupyterhub/blob/73a33ed5fc90f60974141516c84f317f5cebe5e5/jupyterhub/user.py#L859-L860

Jupyterhub startup process calls `jupyterhub.initialize()`, it calls `jupyterhub.init_spawners`:
https://github.com/jupyterhub/jupyterhub/blob/7c5662ee52c98251d6311fd90df6e3542f3d1c71/jupyterhub/app.py#L2873

This method creates `Spawner` instances by fetching all the rows from the database:
https://github.com/jupyterhub/jupyterhub/blob/7c5662ee52c98251d6311fd90df6e3542f3d1c71/jupyterhub/app.py#L2578

`Spawner` constructor calls `spawner._new_spawner()`:
https://github.com/jupyterhub/jupyterhub/blob/73a33ed5fc90f60974141516c84f317f5cebe5e5/jupyterhub/user.py#L241-L248
https://github.com/jupyterhub/jupyterhub/blob/73a33ed5fc90f60974141516c84f317f5cebe5e5/jupyterhub/user.py#L273

Which loads state of each spawner from the database:
https://github.com/jupyterhub/jupyterhub/blob/73a33ed5fc90f60974141516c84f317f5cebe5e5/jupyterhub/user.py#L478

And then all initialized spawners state is checked by calling `spawner.poll()`:
https://github.com/jupyterhub/jupyterhub/blob/7c5662ee52c98251d6311fd90df6e3542f3d1c71/jupyterhub/app.py#L2494

This test is emulating the process of loading spawner state from the database, and using it for starting the pod.

If `c.JupyterHub.cleanup_servers` is `True`, then all spawners' pods are stopped before JupyterHub is shutted down:
https://github.com/jupyterhub/jupyterhub/blob/7c5662ee52c98251d6311fd90df6e3542f3d1c71/jupyterhub/app.py#L2935-L2938

That's why `spawner.stop()` is used here.

Why there is no test with `c.JupyterHub.cleanup_servers=False`?
Well, then JupyterHub is stopped, old spawner instance is gone because the entire JupyterHub process is down. And new instance with the same name and state is created only after JupyterHub process is started again.

So there cannot be 2 spawners with the same parameters, this situation cannot be achieved in JupyterHub, and there is no need to test that.